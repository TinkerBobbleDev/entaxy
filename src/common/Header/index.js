import React from 'react'
import AppBar from 'material-ui/AppBar'
import Toolbar from 'material-ui/Toolbar'
import Typography from 'material-ui/Typography'
import Button from 'material-ui/Button'
import { withStyles } from 'material-ui/styles'

const styles = {
  flex: {
    flex: 1,
  },
};

function Header(props) {
	const { classes } = props
	return (
		<AppBar position="static">
	    <Toolbar>
	      <Typography variant="title" color="inherit" className={classes.flex}>
	        Bean Counter
	      </Typography>
	      <Button color="inherit">Login</Button>
	    </Toolbar>
	  </AppBar>
  )
}

export default withStyles(styles)(Header)