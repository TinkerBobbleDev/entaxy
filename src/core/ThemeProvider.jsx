import React from 'react'
import PropTypes from 'prop-types'
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles'
import CssBaseline from '@material-ui/core/CssBaseline'
import blueGrey from '@material-ui/core/colors/blueGrey'
import '../styles/globals.css'

const theme = createMuiTheme({
  palette: {
    primary: { main: blueGrey[800] },
    secondary: { main: '#1B9CC4' },
    info: {
      text: '#00acc1',
      background: '#00acc11a'
    },
    danger: {
      text: '#721c24',
      background: '#f8d7da66'
    }

  },
  typography: {
    useNextVariants: true
  }
})

const ThemeProvider = ({ children }) => (
  <CssBaseline>
    <MuiThemeProvider theme={theme}>
      {children}
    </MuiThemeProvider>
  </CssBaseline>
)

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired
}

export default ThemeProvider
