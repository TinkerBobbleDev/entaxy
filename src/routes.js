import React from 'react'
import { Route, Router } from 'react-router-dom'
import Taxes from './core/Taxes'

const Routes = ({history}) => (
  <Router history={history}>
    <Route exact path="/taxes" component={Taxes} />
  </Router>
)

export default Routes
