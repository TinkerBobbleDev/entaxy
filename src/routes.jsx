import React from 'react'
import { BrowserRouter, Switch, Route } from 'react-router-dom'
import HandleLogin from './common/HandleLogin'
import Landing from './core/Landing'
import LoadingOverlay from './common/LoadingOverlay'
import Taxes from './core/Taxes'
import Portfolios from './core/Portfolios'
import Transactions from './core/Portfolios/Transactions'
import DataSources from './core/Portfolios/DataSources'

const Routes = () => (
  <div>
    <BrowserRouter basename={process.env.PUBLIC_URL}>
      <Switch>
        <Route exact path="/" component={Landing} />
        <Route exact path="/handle-login" component={HandleLogin} />
        <Route exact path="/taxes" component={Taxes} />
        <Route path="/portfolio" component={Portfolios} />
        <Route path="/transactions" component={Transactions} />
        <Route path="/data-sources" component={DataSources} />
      </Switch>
    </BrowserRouter>
    <LoadingOverlay />
  </div>
)

export default Routes
