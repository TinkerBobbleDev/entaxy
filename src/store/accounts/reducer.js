/* eslint-disable no-case-declarations */
import types from './types'

export const initialState = {
  byId: {}, // {id1: {id: id1, name: 'account1', institution: 'TD'}, id2: {...}}
  byInstitution: {} // {TD: {accountIds: [id1, id2], balance: 100, apiKey: XYZ}, BMO: {...}}
}

export const groupByInstitution = ({ byId, byInstitution }) => {
  // Group sorted accounts by institution
  // results in something like:
  // {
  //   TD: {accountIds: [id1, id2], balance: 100},
  //   BMO: {accountIds: [id3], balance: 200}
  // }
  const groupedAccountIds = Object.values(byId)
    // Sort accounts by name
    .sort((a, b) => (a.name > b.name))
    // Group them by institution
    .reduce((result, account) => {
      const {
        id,
        institution,
        groupId,
        currentBalance
      } = account

      // Merge existing institution data
      const existingInstitutionData = byInstitution[institution] || { groups: {} }
      const currentInstitutionData = (result[institution] || { balance: 0, groups: {} })
      const newInstitutionData = { ...existingInstitutionData, ...currentInstitutionData }

      // Merge existing group data
      const existingGroupData = existingInstitutionData.groups[groupId] || {}
      const currentGroupData = currentInstitutionData.groups[groupId] || {
        id: groupId,
        balance: 0,
        accountIds: []
      }
      const newGroupData = { ...existingGroupData, ...currentGroupData }

      return {
        ...result,
        [institution]: {
          ...newInstitutionData,
          balance: currentInstitutionData.balance + currentBalance.localCurrency,
          groups: {
            ...newInstitutionData.groups,
            [groupId]: {
              type: 'default',
              ...newGroupData,
              balance: currentGroupData.balance + currentBalance.localCurrency,
              accountIds: [...newGroupData.accountIds, id]
            }
          }
        }
      }
    }, {})

  // But now the institutions are not sorted and we lost existing institution data
  return Object.keys(groupedAccountIds)
    // Sort institutions by name
    .sort((a, b) => (a > b))
    // Merge new sorted institutions with existing institution data
    .reduce((result, institution) => ({
      ...result,
      [institution]: {
        ...byInstitution[institution],
        ...groupedAccountIds[institution]
      }
    }), {})
}

let accounts
let institutionData
export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_ACCOUNTS:
      return payload || initialState
    case types.CREATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.UPDATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.DELETE_ACCOUNT:
      const { [payload]: _, ...rest } = state.byId
      return {
        ...state,
        byId: rest
      }
    case types.GROUP_BY_INSTITUTION:
      return {
        ...state,
        byInstitution: groupByInstitution(state)
      }
    case types.CREATE_ACCOUNT_GROUP:
    case types.UPDATE_ACCOUNT_GROUP:
      if (payload === undefined) return state
      const { institution, accountGroup } = payload
      institutionData = state.byInstitution[institution] || { groups: {} }
      return {
        ...state,
        byInstitution: {
          ...state.byInstitution,
          [institution]: {
            ...institutionData,
            groups: {
              ...institutionData.groups,
              [accountGroup.id]: accountGroup
            }
          }
        }
      }
    default:
      return state
  }
}
