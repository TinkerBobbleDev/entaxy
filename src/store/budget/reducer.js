import types from './types'
import budgetCategories from '../../data/budgetCategories'

// https://projects.susielu.com
// http://repec.sowi.unibe.ch/stata/palettes/index.html
const defaultColours = [
  '#1f78b4', '#b2df8a', '#e31a1c',
  '#ff7f00', '#cab2d6', '#a6cee3',
  '#33a02c', '#6a3d9a', '#fb9a99',
  '#fdbf6f', '#ffff99', '#b15928'
]

export const initialState = (() => {
  const colours = {}
  let count = 0
  const categories = Object.keys(budgetCategories).sort().map(category => (
    {
      label: category,
      options: budgetCategories[category].map((subCategory) => {
        const colour = defaultColours[count % defaultColours.length]
        colours[subCategory] = colour
        count += 1
        return { label: subCategory, value: subCategory, colour }
      })
    }
  ))
  return {
    categories,
    colours,
    rules: {} // rules have the format => match: { category: 'cat 1', type: 'exact_match', count: 0 }
  }
})()

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_BUDGET:
      return action.payload || initialState
    case types.CREATE_EXACT_RULE:
      return {
        ...state,
        rules: {
          ...state.rules,
          [action.payload.match]: {
            category: action.payload.category,
            type: 'exact_match',
            count: 0
          }
        }
      }
    case types.DELETE_EXACT_RULE:
      // eslint-disable-next-line no-case-declarations
      const r = Object.keys(state.rules).reduce((result, match) => {
        if (match !== action.payload) {
          return {
            ...result,
            rules: { ...result.rules, [match]: state.rules[match] }
          }
        }
        return result
      }, { ...state, rules: {} })
      return r
    case types.COUNT_RULE_USAGE:
      // reset counters
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(state.rules).reduce((result, match) => (
        {
          ...result,
          rules: {
            ...result.rules,
            [match]: { ...result.rules[match], count: 0 }
          }
        }
      ), state)

      // count the transactions
      for (const transaction of action.payload) {
        if (transaction.description in newState.rules) {
          newState.rules[transaction.description].count += 1
        }
      }
      // remove unused rules
      return Object.keys(newState.rules).reduce((result, match) => {
        if (newState.rules[match].count > 0) {
          return {
            ...result,
            rules: { ...result.rules, [match]: newState.rules[match] }
          }
        }
        return result
      }, { ...newState, rules: {} })
    default:
      return state
  }
}
