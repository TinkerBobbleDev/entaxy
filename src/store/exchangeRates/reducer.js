import { omit } from 'lodash'
import parse from 'date-fns/parse'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_EXCHANGE_RATES:
      return action.payload || initialState
    case types.UPDATE_EXCHANGE_RATES:
      // payload = {
      //   '2019-01-01': { GBP: 1.4, CAD: 1.0 },
      //   '2019-02-02': { GBP: 1.3}
      // }
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(action.payload)
        .reduce((res1, date) => {
          return Object.keys(action.payload[date])
            .reduce((res2, currency) => ({
              ...res2,
              [currency]: {
                ...res2[currency],
                [parse(date).getTime()]: action.payload[date][currency]
              }
            }), res1)
        }, state)
      Object.keys(newState).forEach((currency) => {
        newState[currency].dates = Object.keys(newState[currency])
          .filter(c => c !== 'dates')
          .sort((a, b) => b - a)
      })
      return newState
    case types.DELETE_CURRENCIES:
      return omit(state, action.payload)
    default:
      return state
  }
}
