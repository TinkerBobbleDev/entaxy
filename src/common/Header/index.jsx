import React from 'react'
import PropTypes from 'prop-types'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import { withStyles } from '@material-ui/core/styles'
import Logo from '../Logo/index'
import LoginButton from '../LoginButton'
import TopNav from '../TopNav'

const styles = {
  toolbar: {
    display: 'flex',
    'justify-content': 'space-between'
  }
}

const Header = ({ classes }) => (
  <AppBar position="static">
    <Toolbar className={classes.toolbar}>
      <Logo />
      <TopNav />
      <LoginButton />
    </Toolbar>
  </AppBar>
)

Header.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Header)
