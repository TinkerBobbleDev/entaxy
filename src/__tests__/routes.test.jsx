import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { Provider } from 'react-redux'
import { BrowserRouter, Redirect } from 'react-router-dom'
import store from '../store'
import ThemeProvider from '../core/ThemeProvider'
import { initialState as userInitialState } from '../store/user/reducer'
import { initialState as settingsInitialState } from '../store/settings/reducer'
import { initialState as accountsInitialState } from '../store/accounts/reducer'
import { RoutesComponent } from '../routes'
import Header from '../common/Header'

jest.mock('../common/InstitutionIcon', () => 'InstitutionIcon')

const FakeComponent = () => (<div />)

describe('Routes', () => {
  describe('snapshot', () => {
    it('matches with no accounts', () => {
      const component = renderer.create((
        <Provider store={store}>
          <ThemeProvider>
            <BrowserRouter>
              <RoutesComponent
                user={userInitialState}
                settings={settingsInitialState}
                accounts={accountsInitialState}
                classes={{ }}
              />
            </BrowserRouter>
          </ThemeProvider>
        </Provider>
      ))
      expect(component.toJSON()).toMatchSnapshot()
    })
  })

  describe('Component methods', () => {
    let wrapper
    it('Should redirect if user is not authenticated', () => {
      wrapper = shallow((
        <RoutesComponent
          user={userInitialState}
          settings={settingsInitialState}
          accounts={accountsInitialState}
          classes={{ }}
        />
      ))
      const instance = wrapper.instance()

      expect(instance.props.user.isAuthenticatedWith).toBeNull()
      expect(instance.loginRequired(FakeComponent)()).toEqual(<Redirect to="/" />)
    })

    it('Should render component with no options', () => {
      wrapper = shallow((
        <RoutesComponent
          user={{ isAuthenticatedWith: true }}
          settings={settingsInitialState}
          accounts={accountsInitialState}
          classes={{ }}
        />
      ))
      const instance = wrapper.instance()

      expect(instance.props.user.isAuthenticatedWith).toBeTruthy()
      expect(instance.loginRequired(FakeComponent)()).toEqual((
        <Header><FakeComponent /></Header>
      ))
    })

    it('Should redirect if there are no accounts', () => {
      wrapper = shallow((
        <RoutesComponent
          user={{ isAuthenticatedWith: true }}
          settings={settingsInitialState}
          accounts={accountsInitialState}
          classes={{ }}
        />
      ))
      const instance = wrapper.instance()

      expect(instance.props.user.isAuthenticatedWith).toBeTruthy()
      expect(instance.loginRequired(FakeComponent, { accountRequired: true })()).toEqual((
        <Redirect to="/" />
      ))
    })

    it('Should render component when accounts required', () => {
      wrapper = shallow((
        <RoutesComponent
          user={{ isAuthenticatedWith: true }}
          settings={settingsInitialState}
          accounts={{ byId: [{ id: 1 }] }}
          classes={{ }}
        />
      ))
      const instance = wrapper.instance()

      expect(instance.props.user.isAuthenticatedWith).toBeTruthy()
      expect(instance.loginRequired(FakeComponent, { accountRequired: true })()).toEqual((
        <Header><FakeComponent /></Header>
      ))
    })
  })
})
