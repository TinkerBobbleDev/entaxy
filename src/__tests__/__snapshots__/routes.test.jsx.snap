// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes snapshot matches with no accounts 1`] = `
<div>
  <div
    className="MuiGrid-container-8 LandingComponent-root-1"
  >
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-6-42 LandingComponent-left-2"
    >
      <div
        className="LandingComponent-logo-3"
      >
        Entaxy
        <div
          className="LandingComponent-tagline-4"
        >
          Order from chaos
        </div>
      </div>
      <div
        className="LandingComponent-title-5"
      >
        Your Personal Finance Simple & Private
        <div
          className="LandingComponent-description-6"
        >
          Insights into your finances, without sacrificing your data
        </div>
      </div>
      <div
        className="MuiGrid-container-8"
      >
        <h5
          className="MuiTypography-root-111 MuiTypography-h5-127 MuiTypography-colorSecondary-142"
        >
          Where do you want to store your data?
        </h5>
        <div
          className="MuiGrid-item-9 MuiGrid-grid-xs-6-42"
        >
          <div
            className="MuiPaper-root-147 MuiPaper-elevation1-150 MuiPaper-rounded-148 LandingCardComponent-loggedInContainer-106"
          >
            <h6
              className="MuiTypography-root-111 MuiTypography-subtitle2-130 LandingCardComponent-blockstackTitle-109"
            >
              Anywhere with blockstack
            </h6>
            <span
              className="MuiTypography-root-111 MuiTypography-caption-121 LandingCardComponent-signingDescription-107"
            >
              Keep your data encrypted and decentralized. Learn more about Â 
              <a
                href="https://blockstack.org/what-is-blockstack"
                rel="noopener noreferrer"
                target="_blank"
              >
                Blockstack
              </a>
            </span>
            <button
              className="MuiButtonBase-root-200 MuiButton-root-174 MuiButton-contained-185 MuiButton-containedSecondary-187 MuiButton-raised-188 MuiButton-raisedSecondary-190 LandingCardComponent-signinButton-108"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-175"
              >
                Sign in with Blockstack
              </span>
              <span
                className="MuiTouchRipple-root-214"
              />
            </button>
          </div>
        </div>
        <div
          className="MuiGrid-item-9 MuiGrid-grid-xs-6-42"
        >
          <div
            className="MuiPaper-root-147 MuiPaper-elevation1-150 MuiPaper-rounded-148 LandingCardComponent-loggedInContainer-106"
          >
            <h6
              className="MuiTypography-root-111 MuiTypography-subtitle2-130"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-203 LandingCardComponent-browserIcon-110"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <defs>
                  <path
                    d="M0 0h24v24H0V0z"
                    id="a"
                  />
                </defs>
                <path
                  d="M20 18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2H0v2h24v-2h-4zM4 6h16v10H4V6z"
                />
              </svg>
              Locally in your browser
            </h6>
            <span
              className="MuiTypography-root-111 MuiTypography-caption-121 LandingCardComponent-signingDescription-107"
            >
              Your data will be stored temporarily and it be removed once you close your browser
            </span>
            <button
              className="MuiButtonBase-root-200 MuiButton-root-174 MuiButton-contained-185 MuiButton-containedSecondary-187 MuiButton-raised-188 MuiButton-raisedSecondary-190 LandingCardComponent-signinButton-108"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-175"
              >
                Take it for a test drive
              </span>
              <span
                className="MuiTouchRipple-root-214"
              />
            </button>
          </div>
        </div>
      </div>
      <div>
        <p
          className="MuiTypography-root-111 MuiTypography-body2-119"
        >
          Entaxy is free and you get to keep your data.
        </p>
        <p
          className="MuiTypography-root-111 MuiTypography-body2-119"
        >
          That's right, we don't store your data in a big database so we don't need to convince you to trust us.
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-item-9 MuiGrid-grid-xs-6-42 LandingComponent-right-7"
    />
  </div>
</div>
`;
